x/mongo/driver/operation.go

```diff
@@ -470,23 +524,25 @@ func (op Operation) Execute(ctx context.Context, scratch []byte) error {
 				tt.Labels = append(tt.Labels, RetryableWriteError)
 			}
 
-			if retryable && retryableErr && retries != 0 {
-				retries--
-				original = err
-				conn.Close() // Avoid leaking the connection.
-				srvr, conn, err = op.getServerAndConnection(ctx)
-				if err != nil || conn == nil || !op.retryable(conn.Description()) {
-					if conn != nil {
-						conn.Close()
-					}
-					return original
-				}
-				defer conn.Close() // Avoid leaking the new connection.
+			// If retries are supported for the current operation on the first server description,
+			// the error is considered retryable, and there are retries remaining (negative retries
+			// means retry indefinitely), then retry the operation.
+			if retrySupported && retryableErr && retries != 0 {
 				if op.Client != nil && op.Client.Committing {
 					// Apply majority write concern for retries
 					op.Client.UpdateCommitTransactionWriteConcern()
 					op.WriteConcern = op.Client.CurrentWc
 				}
+				retries--
+				prevErr = tt
+				// If we got a connection, close it immediately to release pool resources for
+				// subsequent retries.
+				if conn != nil {
+					conn.Close()
+				}
+				// Set the server and connection to nil to request a new server and connection.
+				srvr = nil
+				conn = nil
```

Comment:
Given that this exact code is duplicated in the `WriteCommandError` and `Error` cases, can we write a closure within this function to handle this logic? Suggesting a closure rather than a separate helper function as it'll be easier to do all of the state management like setting `srvr` and `conn` to nil in a closure that can directly access those variables.

Reply:
Sounds like a good idea, will add a closure that resets the variable states for retry.

Reply:
Added the closure `resetForRetry` that captures most of the common code for resetting the retry loop variables.